### ISSUE TITLE
Normalization through machine integer libraries breaks abstraction

### ISSUE THREAD

**Issue by tchajed** [FStarLang/FStar#1207]
Here's an example of code that doesn't extract correctly due to reductions that uncover the definitions in `UInt32.fst` (and by analogy the rest of the machine integer sizes). I've inlined the normalization tactic to avoid FStarLang/kremlin#54.

```
module Int_reduce

open FStar.Tactics

module UInt32 = FStar.UInt32

// reminder that this is the desugaring
let _ = assert (3ul == UInt32.uint_to_t 3)

// extracts
let ret_3 (_:unit) : UInt32.t = 3ul

// does not extract
let ret_3_norm (_:unit) : UInt32.t =
  synth_by_tactic (dup;;
    exact (quote 3ul);;
    norm [Delta];;
    trefl)

// extracts
let compare_ints (x y:UInt32.t) : bool = UInt32.lt x y

// does not extract
let compare_ints_norm (x y:UInt32.t) : bool =
  synth_by_tactic (dup;;
    exact (quote (UInt32.lt x y));;
    norm [Delta];;
    trefl)
```

I'm specifically talking about KreMLin extraction above, but the `ret_3_norm` example is broken even in OCaml because it reveals the `UInt32.Mk` constructor, which is not realized in `FStar_UInt32.ml` (and can't be because it's not a constructor in OCaml and we can't start with a capital letter).

This immediate issue can be fixed by making `uint_to_t` abstract. The comparison operators need specifications that reveal the current behavior, then they can be made abstract as well. The proper fix I'd like to see is that all of the machine integer libraries are properly abstracted by interfaces.

**Comment by nikswamy** [comment]
I've been working on this in the nik_bvtac branch. Expect a properly abstracted version of the machine integers to land in master soon.

**Comment by nikswamy** [comment]
Closed, since 58d0d92
